### ISSUE TITLE
Question: is there a supported/suggestion server side Pagination implementation for Connection Types

### ISSUE THREAD

**Issue by joelwarrington-jobber** [rmosolgo/graphql-ruby#3576]
Hey @rmosolgo  👋 ,

Is there any supported/suggested method for handling Pagination server side when querying for ActiveRecord while limiting the number of items returned?

I know of  `max_page_size` and `default_max_page_size` however this requires you to still return the entire collection to be able to paginate and also get accurate item totals. Are there any out-of-the-box implementations which are recommended?

Example:

```ruby
field :dogs, resolver: Resolvers::DogsResolver


def DogsResolver
  type Types::DogType.connection_type, null: false

  def resolve
    # assume there are a lot of dogs, and the limit number would be something manageable (like 100)
    Dog.all.limit(1)
  end
end
```

```gql
query DogsQuery {
  dogs {
    pageInfo {
      startCursor
      hasPreviousPage
      hasNextPage
      endCursor
    }
    totalCount
  }
}
```

```json
{
  "data": {
    "dogs": {
      "pageInfo": {
        "startCursor": "MQ",
        "hasPreviousPage": false,
        "hasNextPage": false,
        "endCursor": "MQ"
      },
      "totalCount": 1
    }
  }
}
```

**Comment by joelwarrington-jobber** [comment]
@rmosolgo, my method to confirm may have been flawed, when adding a `binding.pry`, and getting the `.to_sql` of the query it didn't include a limit. Does the `max_page_size` limit ActiveRecord queries somehow, and will this still provide the `totalCount` that is accurate (even if limited to something below that)

**Comment by rmosolgo** [comment]
Yes, I expect `max_page_size` to limit the query to fetch objects while `totalCount` returns an unbounded count.

**Comment by joelwarrington-jobber** [comment]
Awesome, thanks for the clarification, will do a bit more confirmation on my side and add any findings here so that if someone in the future finds this they also can get clarification. 😊

**Comment by joelwarrington-jobber** [comment]
**🔔  closing issue, as working as expected**

---

The following comment includes some investigation as to why I saw `limit` not being applied to the query when using `.connection_type` and `default_max_page_size`, but in short, limiting and database pagination works as expected when using the built in connection type and max page size for ActiveRecord.

#### Why did the query not show `limit` being correctly applied?

When running a Rails Console or when in a `binding.pry`, `ActiveRecord_AssociationRelation` class is evaluated right away and prints out the result, but in practice isn't evaluated right away in the resolve method. That means that doing something like the following:

```ruby
field :dogs, Types::DogType

def dogs
  result = Dog.all
  binding.pry
  result
end
```

and checking `result` will actually evaluate the result, where as if you don't include this, `ActiveRecord_AssociationRelation` is returned. Which gets appropriately limited here: https://github.com/rmosolgo/graphql-ruby/blob/7cebf9bd0b465f88f85a01a32e966b2db4180a3f/lib/graphql/pagination/relation_connection.rb#L104

#### My `max_page_size` (or default) is being ignored?!?

If you limit the resolver/query yourself (example following), pagination will still work using the cursors, however the items returned will be limited to a page size of your `limit` within the resolver. Also, if you've implemented a `total_count` field on your base connection and don't unscope `limit` for the `.count` query, you'll get deceptive results

```ruby
field :dogs, Types::DogType, max_page_size: 50

def dogs
  # this limit will supersede the `max_page_size` or `default_max_page_size` from my investigation
  Dog.all.limit(5)
end
```
